consensus/core/src/transactions_client.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::block::Transaction;
+use crate::context::Context;
+use mysten_metrics::metered_channel;
+use mysten_metrics::metered_channel::channel_with_total;
+use std::sync::Arc;
+use tap::tap::TapFallible;
+use thiserror::Error;
+use tracing::warn;
+
+/// Maximum number of transactions to be fetched per request of `next`
+const MAX_FETCHED_TRANSACTIONS: usize = 100;
+/// The maximum number of transactions pending to the queue to be pulled for block proposal
+const MAX_PENDING_TRANSACTIONS: usize = 2_000;
+
+/// The TransactionsConsumer is responsible for fetching the next transactions to be included for the block proposals.
+/// The transactions are submitted to a channel which is shared between the TransactionsConsumer and the TransactionsClient
+/// and are pulled every time the `next` method is called.
+#[allow(dead_code)]
+pub(crate) struct TransactionsConsumer {
+    tx_receiver: metered_channel::Receiver<Transaction>,
+    max_fetched_per_request: usize,
+}
+
+#[allow(dead_code)]
+impl TransactionsConsumer {
+    pub(crate) fn new(tx_receiver: metered_channel::Receiver<Transaction>) -> Self {
+        Self {
+            tx_receiver,
+            max_fetched_per_request: MAX_FETCHED_TRANSACTIONS,
+        }
+    }
+
+    pub(crate) fn with_max_fetched_per_request(mut self, max_fetched_per_request: usize) -> Self {
+        self.max_fetched_per_request = max_fetched_per_request;
+        self
+    }
+
+    // Attempts to fetch the next transactions that have been submitted for sequence.
+    pub(crate) fn next(&mut self) -> Vec<Transaction> {
+        let mut transactions = Vec::new();
+        while let Ok(transaction) = self.tx_receiver.try_recv() {
+            transactions.push(transaction);
+
+            if transactions.len() >= self.max_fetched_per_request {
```

Comment:
nit: should we just keep a running counter instead of checking `len()` each iteration?

Reply:
we could do, but I don't think there is going to be much of a difference, as the `len()` [is using a cached value](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#2107) as well ?